optuna/storages/rdb/alembic/versions/v1.2.0.b.py

```diff
@@ -0,0 +1,91 @@
+"""empty message
+
+Revision ID: v1.2.0.b
+Revises: v1.2.0.a
+Create Date: 2020-02-14 16:23:04.800808
+
+"""
+import json
+
+from alembic import op
+import sqlalchemy as sa
+
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy import orm
+
+# revision identifiers, used by Alembic.
+revision = 'v1.2.0.b'
+down_revision = 'v1.2.0.a'
+branch_labels = None
+depends_on = None
+
+# Model definition
+MAX_INDEXED_STRING_LENGTH = 512
+MAX_STRING_LENGTH = 2048
+BaseModel = declarative_base()
+
+
+class TrialModel(BaseModel):
+    __tablename__ = 'trials'
+    trial_id = sa.Column(sa.Integer, primary_key=True)
+    number = sa.Column(sa.Integer)
+
+
+class TrialSystemAttributeModel(BaseModel):
+    __tablename__ = 'trial_system_attributes'
+    trial_system_attribute_id = sa.Column(sa.Integer, primary_key=True)
+    trial_id = sa.Column(sa.Integer, sa.ForeignKey('trials.trial_id'))
+    key = sa.Column(sa.String(MAX_INDEXED_STRING_LENGTH))
+    value_json = sa.Column(sa.String(MAX_STRING_LENGTH))
+
+
+def upgrade():
+    bind = op.get_bind()
+    session = orm.Session(bind=bind)
+
+    with op.batch_alter_table("trials") as batch_op:
+        batch_op.add_column(sa.Column('number', sa.Integer(), nullable=True, default=-1))
+        batch_op.create_index('ix_study_trial_number', ['study_id', 'number']),
+
+    try:
+        number_records = session.query(TrialSystemAttributeModel) \
+            .filter(TrialSystemAttributeModel.key == '_number').all()
+        mapping = [{'trial_id': r.trial_id, 'number': json.loads(r.value_json)}
+                   for r in number_records]
+        session.bulk_update_mappings(TrialModel, mapping)
+
+        session.query(TrialSystemAttributeModel.key == '_number') \
+            .delete(synchronize_session=False)
+        session.commit()
+    except SQLAlchemyError as e:
+        session.rollback()
+        raise e
+    finally:
+        session.close()
+
+
+def downgrade():
```

Comment:
I think optuna only provides `upgrade` officially (i.e., `optuna storage upgrade`) and does not support `downgrade`. So, I could not come up with appropriate ways to validate this function.
Is it sufficient if I execute the `alembic downgrade` in `optuna/storages/rdb`?
If you have any comments, please let me know.

Reply:
Yes. I checked the behavior of this function by `alembic downgrade` command.

1. Run a something example with `sqlite:///alembic.db`.
1. Put a `alembic.db` file to `optuna/storages/rdb`.
1. Run `alembic downgrade v1.2.0.a`.

Reply:
Maybe this information is useful for future developers of `RDBStorage`. Thanks.